service/history/taskProcessingJob.go

```diff
@@ -0,0 +1,646 @@
+) domainFilter {
+	filter := domainFilter{
+		domainIDs:    make(map[string]struct{}),
+		invertResult: inverted,
+	}
+	for _, domainID := range domainIDs {
+		filter.domainIDs[domainID] = struct{}{}
+	}
+	return filter
+}
+
+func (f *domainFilter) filter(domainID string) bool {
+	_, ok := f.domainIDs[domainID]
+	if f.invertResult {
+		ok = !ok
+	}
+	return ok
+}
+
+func (f *domainFilter) include(domainIDs []string) domainFilter {
+	filter := f.copy()
+	for _, domainID := range domainIDs {
+		if !filter.invertResult {
+			filter.domainIDs[domainID] = struct{}{}
+		} else {
+			delete(filter.domainIDs, domainID)
+		}
+	}
+	return filter
+}
+
+func (f *domainFilter) exclude(domainIDs []string) domainFilter {
+	filter := f.copy()
+	for _, domainID := range domainIDs {
+		if !filter.invertResult {
+			delete(filter.domainIDs, domainID)
+		} else {
+			filter.domainIDs[domainID] = struct{}{}
+		}
+	}
+	return filter
+}
+
+func (f *domainFilter) merge(f2 domainFilter) domainFilter {
+	if !f.invertResult && !f2.invertResult {
+		// union the domains
+		filter := f.copy()
+		for domainID := range f2.domainIDs {
+			filter.domainIDs[domainID] = struct{}{}
+		}
+		return filter
+	}
+
+	if f.invertResult && f2.invertResult {
+		// intersect the domains
+		filter := domainFilter{
+			domainIDs:    make(map[string]struct{}),
+			invertResult: true,
+		}
+		for domainID := range f.domainIDs {
+			if _, ok := f2.domainIDs[domainID]; ok {
+				filter.domainIDs[domainID] = struct{}{}
+			}
+		}
+		return filter
+	}
+
+	var filter domainFilter
+	var includeDomainIDs map[string]struct{}
+	if f.invertResult {
+		filter = f.copy()
+		includeDomainIDs = f2.domainIDs
+	} else {
+		filter = f2.copy()
+		includeDomainIDs = f.domainIDs
+	}
+
+	for domainID := range includeDomainIDs {
+		delete(filter.domainIDs, domainID)
+	}
+
+	return filter
+}
+
+func (f *domainFilter) copy() domainFilter {
+	domainIDs := []string{}
+	for domainID := range f.domainIDs {
+		domainIDs = append(domainIDs, domainID)
+	}
+	return newDomainFilter(domainIDs, f.invertResult)
+}
+
+func newBaseSplitPolicy(
+	maxJobLevel int,
+	totalPendingTaskThreshold int,
+	perDomainPendingTaskThreshold int,
+) taskProcessingJobSplitPolicy {
+	return &baseSplitPolicy{
+		maxJobLevel:                   maxJobLevel,
+		totalPendingTaskThreshold:     totalPendingTaskThreshold,
+		perDomainPendingTaskThreshold: perDomainPendingTaskThreshold,
+	}
+}
+
+func (p *baseSplitPolicy) Evaluate(
+	pendingTaskStats map[string]int,
+	currentLevel int,
+) map[string]int {
+	if currentLevel >= p.maxJobLevel {
+		return nil
+	}
+
+	totalPendingTaskNum := 0
+	domainIDToNewLevel := make(map[string]int)
+	for domainID, numPendingTasks := range pendingTaskStats {
+		totalPendingTaskNum += numPendingTasks
+		if numPendingTasks > p.perDomainPendingTaskThreshold {
+			domainIDToNewLevel[domainID] = currentLevel + 1
+		}
+	}
+
+	if totalPendingTaskNum < p.totalPendingTaskThreshold {
+		return nil
+	}
+	return domainIDToNewLevel
```

Comment:
so if `if totalPendingTaskNum < p.totalPendingTaskThreshold` then even if # of domain tasks exceeds `perDomainPendingTaskThreshold` nothing is done? then why having `totalPendingTaskThreshold`?